waldo/upload.go

```diff
@@ -190,58 +268,124 @@ func (ua *UploadAction) makeArgs(buildPath, uploadToken string) []string {
 		args = append(args, "--verbose")
 	}
 
-	if ua.options.Version {
-		args = append(args, "--version")
-	}
-
 	return args
 }
 
-func (ua *UploadAction) mungeArgs(ud *data.UserData, recipe *data.Recipe) ([]string, string, error) {
-	uploadToken := ua.options.UploadToken
+func (ua *UploadAction) processOptions(ciMode bool) (*data.UserData, *data.Recipe, error) {
+	isRecipe := ua.isRecipeTarget()
 
-	// if len(uploadToken) == 0 {
-	// 	uploadToken = recipe.UploadToken
-	// }
+	var err error
 
-	if len(uploadToken) == 0 {
-		uploadToken = os.Getenv("WALDO_UPLOAD_TOKEN")
+	//
+	// - An app token (formerly known as an “upload token”) is _allowed_
+	//   only if the target is _not_ a recipe name.
+	// - An app token is _required_ if and only if we are in CI Mode.
+	//
+	ua.appToken, err = ua.detectAppToken(ciMode, !isRecipe)
+
+	if err != nil {
+		return nil, nil, err
 	}
 
-	err := data.ValidateUploadToken(uploadToken)
+	//
+	// In CI mode, the target _must not_ be a recipe name:
+	//
+	if ciMode {
+		_, err = ua.detectRecipeName(false, false)
 
-	if err != nil {
-		return nil, "", err
+		return nil, nil, err
 	}
 
-	target := ua.options.Target
+	//
+	// Otherwise, the target _may_ be a recipe name iff an app token has _not_
```

Comment:
Same remark but for another reason.

If I set a WALDO_APP_TOKEN in my env to test the CI upload path then call the command with a valid recipe name it will fail.

And i'll need to remember that I put the value in my env somewhere (as the error message will point to my command line) and how to remove it from there

Reply:
Do you have a solution in mind that does not break existing customers?

Reply:
Yes, not sure my remark was precise enough but it's actually a continuation of the "need to convince larry" one and is about:

>  In CI mode, the target _must not_ be a recipe name:

Essentially let's not error when an app token was provided with a recipe. Or at minimum let's distinguish between command line and env and don't fail when an app token come from the env and a receipe name is provided on the cmdline.

Reply:
Are you suggesting that in the case where `WALDO_UPLOAD_TOKEN` is specified in the environment _and_ a recipe name is provided, that the envirnoment variable should be ignored?